### PR TITLE
Editorial: Add prose about constructing a request

### PR DESCRIPTION
**Pull request by noamr**
This intends to be a summary of informative notes in the #Requests section, into a concise "starter" guide to populating a request, which is usually the tricky bit of using Fetch.

Requested by authors of other specs, see for example: https://github.com/w3ctag/design-principles/issues/238 https://github.com/fedidcg/FedCM/issues/320

***
<a href="https://whatpr.org/fetch/1585.html" title="Last updated on Mar 2, 2023, 1:03 PM UTC (dd89623)">Preview</a> | <a href="https://whatpr.org/fetch/1585/a5560d2...dd89623.html" title="Last updated on Mar 2, 2023, 1:03 PM UTC (dd89623)">Diff</a>

### CODE REVIEW

**Comment by noamr** [comment]
@annevk I'm sure some things here can be made more accurate but I thought I'd take a first stab as I feel this is needed...

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8403]: It might work better to write this as a [`<dl class="switch">`](https://w3c.github.io/tr-design/src/README#switch) with the options looking like 
```html
<dt>If the request's URL should be constrained to the same origin as its client's origin
<dd>Set [=request/mode=] to [=same origin=]
```

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8409]: Maybe don't even mention these. We could include something like "Other request modes can only be used by the feature they're named after." to head off questions.

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8417]: I'd give each of these a paragraph or an item in a [`<dl class="switch">`](https://w3c.github.io/tr-design/src/README#switch). Like 
> If you need to verify the received data hashes to a particular value, set [=request/integrity metadata=].

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8383]: A high-level comment for the whole section, which I've elaborated on in several of the specific comments:

I think we want this section to be instructions to the authors of calling specifications -- a how-to guide. It might make sense to also improve the descriptive text in https://fetch.spec.whatwg.org/#requests, but that's not the primary hole this section needs to fill.

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8393]: These probably belong first:

> Start by setting the HTTP [=request/method=] and target [=request/URL=] as described in [[HTTP]]. If your `POST` or `PUT` request needs a body, you can assign or stream it to [=request/body=].

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8420]: Perhaps:

> Set your request's [=request/destination=] according to how your specification is going to use the response. If you need a new destination, ...

This paragraph should link to the note that's under https://fetch.spec.whatwg.org/#request-destination-script-like that describes what each of the destinations and initiator types are for.

I think people are required to set their destination, in which case, this paragraph should appear before the optional parameters.

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8426]: Split this paragraph so people notice that they also need to set the initiator type. Do they always need to set it, or just if they want their resource to participate in [[resource-timing]]? Is the note under https://w3c.github.io/resource-timing/#dom-performanceresourcetiming-initiatortype the best description of how to set this?

**Review by jyasskin — COMMENTED** [review]
Here are my thoughts about how this should work, but you should prioritize opinions from people who've had trouble calling into fetch, like @tabatkins and @npm1.

**Comment by jyasskin** [comment]
Also, thanks a bunch for starting this section!

File: fetch.bs
Comment (npm1) [inline fetch.bs:8389]: nit: initiated (typo)

File: fetch.bs
Comment (npm1) [reply fetch.bs:8383]: Agree that it would be good to format this more as a how-to, and in particular have more bullet lists (and mandatory parameters first and common optional parameters after). A hard part about using fetch is how to invoke it and continue processing the response after it's done. Based on the title of the section it would be out of scope but wanted to point out the gap there as well. There are not many good examples of using fetch in a nontrivial way where the result is 'awaited' by the algorithm, which then does something with the response.

Comment (noamr) [reply fetch.bs:8383]: I focused on requests first because the callback part is already kinda covered [in the subsequent bit](https://fetch.spec.whatwg.org/#fetch-elsewhere-fetch). Is something missing in that section?

Comment (jyasskin) [reply fetch.bs:8383]: I think that subsequent section could use an example of how to use it. I think a lot of people expect to be able to say: 
1. In parallel:
   1. fetch the request.
   2. Wait for the fetch to finish.
   3. Use the response.

And with the setup of the *processResponse* algorithms, that just doesn't work. You have to do something more like

1. Fetch the request, with the following algorithms as arguments:
   <dl><dt>processResponse<dd>Steps
   <dt>processResponseConsumeBody<dd>Other Steps

It might also be helpful to rewrite that section to map from a spec's goals about how it'll handle the response, to the set of algorithm arguments it ought to pass.

But +100 for doing all that in a separate PR. Writing this section is a great first step.

Comment (noamr) [reply fetch.bs:8383]: Gotcha, this makes sense. I will get to improving the callback section once we finish with the request section.
I also intend to add a section about the different cross-origin protections (CORS, TAO, CORB, CSP) with a short guide on which to use for what purpose.

File: fetch.bs
Comment (noamr) [reply fetch.bs]: @jyasskin @npm1 revised based on your comments, I think it's much better now. WDYT?

**Review by npm1 — COMMENTED** [review]
Looks good to me!

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8425]: I wanted to suggest
```suggestion
<p>Figure out if your fetch needs to be reported to [[RESOURCE-TIMING inline]], and with which
```
to get a nice link to the spec, but it gives some ugly extra "Level 2" text as part of the full spec title. (https://github.com/tabatkins/bikeshed/issues/2441) Up to you whether you think it works better anyway.

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8447]: I think this is
```suggestion
 <dd><p>If the site's <cite>Content Security Policy</cite> specifies a <a grammar>nonce-source</a>
 for this <a for=/>request</a>'s <a for=request>destination</a>, any request without a
 [[CSP#match-nonce-to-source-list|matching]] value in this field will be blocked.
```

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8426]:
```suggestion
<p>Figure out if your fetch needs to be reported to <cite>Resource Timing</cite>. If so, set
the <a for=/>request</a>'s <a for=request>initiator type</a> to the value you want to appear
in <code>{{PerformanceResourceTiming}}.{{PerformanceResourceTiming/initiatorType}}</code>.
By passing an <a for=request>initiator type</a> to the
```
(The choice of `{{}}` or `<a idl>` actually affects the styling here.)

**Review by jyasskin — APPROVED** [review]
Thanks! I think @annevk and @domenic should decide how they want this section to direct feature authors to start discussions with them. Do they want to enable Github Discussions here, use Issues, or something else? Beyond that, all of my comments here are optional, and I don't expect you to wait for me once the editors approve the PR.

File: fetch.bs
Comment (noamr) [reply fetch.bs:8402]: The idea is to link "platform object" with a term that people who write other APIs are familiar with

File: fetch.bs
Comment (noamr) [reply fetch.bs:8405]: Can you give examples outside of FedCM? Is this mostly about UA-initiated fetches that are not web-exposed in themselves?

Comment (jyasskin) [reply fetch.bs:8405]: I believe these are all web-exposed:
* https://wicg.github.io/attribution-reporting-api/#create-a-report-request
* https://wicg.github.io/background-fetch/#background-fetch-manager-fetch
* https://w3c.github.io/reporting/#try-delivery

File: fetch.bs
Comment (jyasskin) [inline fetch.bs:8403]: Given https://github.com/whatwg/fetch/pull/1585/files#r1062124820, I'd suggest:

```suggestion
<a>environment settings object</a> you're operating in. Web-exposed APIs are generally defined with
the [[WEBIDL inline]], for which every object that implements an <a>interface</a> has a
<a>relevant settings object</a> you can use. For example, a <a for=/>request</a> associated with a DOM
```

File: fetch.bs
Comment (noamr) [reply fetch.bs:8405]: > I believe these are all web-exposed:
> 
> * https://wicg.github.io/attribution-reporting-api/#create-a-report-request

Seems like the fetch itself is not web exposed?

> 
> * https://wicg.github.io/background-fetch/#background-fetch-manager-fetch

This is a JS API with an IDL interface, which has a relevant settings object.

> 
> * https://w3c.github.io/reporting/#try-delivery

Seems like this ends up also doing something with a JS observer, which has a relevant settings object

Comment (jyasskin) [reply fetch.bs:8405]: For the first, I might be understanding something different by "web-exposed": the result doesn't get reported to Javascript running in a web page, but developers building websites will have their designs broken if it changes.

For the second, the relevant settings object's window may have closed by the time the fetch is sent, which means it can't use that as its *client*. Maybe the right answer is to start a Service Worker to exist while the fetch is running, so that can be the *client*, but the feature's designers need someone to tell them that, and help work through the tradeoffs.

For the third, the JS observer isn't really important, but there is a global available in the [algorithm's caller](https://w3c.github.io/reporting/#send-reports), so the use of `null` is just a bug. But the developers *thought* they didn't have one, maybe because the algorithm originally ran after the originating client closed. They should have a way to ask what to do, so someone knowledgeable can find them the right client.

Comment (noamr) [reply fetch.bs:8405]: Totally agree that documentation for this is lacking. Let's fix it step by step.
btw client is not mandatory. Instead of a client you can populate the request's origin and policy container, and respond to callbacks on a parallel queue (some anonymous new thread). For some of these features maybe that's the right approach and we should address this in this documentation.

**Comment by noamr** [comment]
@annevk want to have a go at reviewing this?

File: fetch.bs
Comment (domenic) [inline fetch.bs:1700]: I would either make this `<dfn>destination table</dfn>`, or not make it a `<dfn>` and instead just add an `id=""` to the note. The latter seems a bit cleaner, as having a definition inside a non-normative section is weird.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8398]: "you can assign or stream it" is confusing. How would you stream it?

I think the most common pattern here will be to use a byte sequence as the body, using the helper we recently defined: https://whatpr.org/fetch/1585.html#byte-sequence-as-a-body . I'd suggest referencing that as the primary method, and mentioning you can construct a new [=body=] for advanced cases.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8401]:
```suggestion
<a>destination table</a>. <a for=request>Destinations</a> affect
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8403]:
```suggestion
[:Sec-Fetch-Dest:] header, so they are much more than informative metadata. If a new
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8406]: Usually we do these as `<a href="https://github.com/whatwg/fetch/issues/new?title=What%20destination%20should%20my%20feature%20use">file an issue</a> to discuss`.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8414]:
```suggestion
web-exposed by JavaScript, HTML, CSS, or other {{Document}} subresources should have a
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8422]:
```suggestion
advanced cases, make sure the details of how your fetch handles <cite>Content Security Policy</cite>
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8432]: Similar suggestion for issue-filing markup.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8439]: Don't you actually just set the header list? Since header list is, as you recently explained to me, more like "extra header list"?

File: fetch.bs
Comment (domenic) [inline fetch.bs:8452]: I think very few fetches will need this. (Only ones initiated by API surfaces which have an explicit place for web developers to put the integrity, basically.) Not sure if it's worth mentioning.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8461]: Ditto.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8466]: Use bikeshed auto-links for the section here, if possible.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8467]: I think service-workers mode often needs configuring for the type of "exotic" fetches we've seen recently. Maybe mention it in the section on null-client fetches.

Similarly such fetches will probably need to set the referrer.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8468]: Cache mode and redirect mode are other parameters I sometimes see changed.

Comment (noamr) [reply fetch.bs:8468]: Does someone use redirect-mode other than navigation & service-workers? Seems pretty exotic

**Comment by noamr** [comment]
Thanks @domenic, I addressed your notes but seems like builds are failing (CSSWG API is down?)

File: fetch.bs
Comment (domenic) [reply fetch.bs:8457]: I see a few places using it to prohibit redirects: https://dontcallmedom.github.io/webdex/r.html#redirect%20mode%40%40request%40dfn

Comment (noamr) [reply fetch.bs:8460]: OK, added a short note about prohibiting redirects ("manual" is the exotic one I guess)

**Comment by noamr** [comment]
> Thanks @domenic, I addressed your notes but seems like builds are failing (CSSWG API is down?)

Any other comments?
Thanks 🙏🏻

File: fetch.bs
Comment (domenic) [inline fetch.bs:1700]: This entire paragraph's diff seems unnecessary, and mostly just inserts a typo.

File: fetch.bs
Comment (domenic) [inline fetch.bs:8413]:
```suggestion
Web IDL, for which every object that implements an <a>interface</a> has a
```

File: fetch.bs
Comment (domenic) [reply fetch.bs:8417]: Still missing this

File: fetch.bs
Comment (domenic) [inline fetch.bs:8426]:
```suggestion
details of how your fetch handles <cite>Content Security Place</cite> and
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8427]:
```suggestion
<a for=/>referrer policy</a> are fleshed out. Also make sure you handle concurrency, as callbacks (see [[#fetch-elsewhere-fetch]])
```

File: fetch.bs
Comment (domenic) [inline fetch.bs:8456]:
```suggestion
<p>Determine whether you want your request to support redirects. If you don't, set its
```

**Review by domenic — COMMENTED** [review]
Still some nits.

File: fetch.bs
Comment (noamr) [reply fetch.bs]: Fixed

**Comment by noamr** [comment]
> Still some nits.

All fix... Another review pass?

**Comment by noamr** [comment]
Ping? @domenic

**Review by domenic — APPROVED** [review]
LGTM, sorry for missing the previous ping!

**Review by annevk — APPROVED** [review]

**Pull request merged**
